File: Leetcode/RemoveNthNodeFromEndofList.cpp
/**
* Definition for singly-linked list.
* struct ListNode {
*     int val;
*     ListNode *next;
*     ListNode(int x) : val(x), next(NULL) {}
* };
*/
class Solution {
public:
	int cntFromEnd = 0;
	ListNode* removeNthFromEnd(ListNode* head, int n) {
		ListNode* res;
		if (head != NULL){
			head->next = removeNthFromEnd(head->next, n);
			++cntFromEnd;
			if (cntFromEnd == n){
				res = head->next;
			}
			else{
				res = head;
			}
		}
		else{
			res = head;
		}
		return res;
	}
};

File: Leetcode/RemoveNthNodeFromEndofList.rs
// https://leetcode.com/problems/remove-nth-node-from-end-of-list

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }

struct Solution {}

impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, mut n: i32) -> Option<Box<ListNode>> {
        Solution::tail(head, &mut n)
    }

    fn tail(cur: Option<Box<ListNode>>, n: &mut i32) -> Option<Box<ListNode>> {
        if let Some(mut cur_) = cur {
            cur_.next = Solution::tail(cur_.next, n);
            *n -= 1;

            if n == &0 {
                cur_.next
            } else {
                Some(cur_)
            }
        } else {
            cur
        }
    }
}
